design-proposals/flavors-and-preferences.md

```diff
@@ -0,0 +1,1023 @@
+  * ioThreadsPolicy
+    * `VirtualMachineFlavorSpec.ioThreadsPolicy`
+  * [launchSecurity](http://kubevirt.io/api-reference/main/definitions.html#_v1_launchsecurity)
+    * [sev](http://kubevirt.io/api-reference/main/definitions.html#_v1_sev)
+      * `VirtualMachineFlavorSpec.LaunchSecurity.SEV`
+  * [Machine](http://kubevirt.io/api-reference/main/definitions.html#_v1_machine)
+    * type
+      * `VirtualMachinePreferenceSpec.Machine.PreferredMachineType`
+  * [Memory](http://kubevirt.io/api-reference/main/definitions.html#_v1_memory)
+    * [guest](http://kubevirt.io/api-reference/main/definitions.html#_k8s_io_apimachinery_pkg_api_resource_quantity) 
+      * `VirtualMachineFlavorSpec.Memory.Guest`
+    * [hugepages](http://kubevirt.io/api-reference/main/definitions.html#_v1_hugepages)
+      * `VirtualMachineFlavorSpec.Memory.HugePages`
+
+## Types
+
+### `VirtualMachineFlavorSpec`
+
+```go
+// VirtualMachineFlavorSpec
+//
+// +k8s:openapi-gen=true1
+type VirtualMachineFlavorSpec struct {
+	CPU *CPUFlavor `json:"cpu"`
+
+	Memory *MemoryFlavor `json:"memory"`
+
+	// Optionally defines any GPU devices associated with the flavor.
+	//
+	// +optional
+	// +listType=atomic
+	GPUs []v1.GPU `json:"gpus,omitempty"`
+
+	// Optionally defines any HostDevices associated with the flavor.
+	//
+	// +optional
+	// +listType=atomic
+	HostDevices []v1.HostDevice `json:"hostDevices,omitempty"`
+
+	// Optionally defines the IOThreadsPolicy to be used by the flavor.
+	//
+	// +optional
+	IOThreadsPolicy *v1.IOThreadsPolicy `json:"ioThreadsPolicy,omitempty"`
+
+	// Optionally defines the LaunchSecurity to be used by the flavor.
+	//
+	// +optional
+	LaunchSecurity *v1.LaunchSecurity `json:"launchSecurity,omitempty"`
+}
+```
+
+### `CPUFlavor`
+
+```go
+// CPUFlavor
+//
+// +k8s:openapi-gen=true
+type CPUFlavor struct {
+
+	// Number of vCPUs to expose to the guest.
+	// The resulting CPU topology being derived from the optional PreferredCPUTopology attribute of CPUPreferences.
+	Guest uint32 `json:"guest"`
+
+	// Model specifies the CPU model inside the VMI.
+	// List of available models https://github.com/libvirt/libvirt/tree/master/src/cpu_map.
+	// It is possible to specify special cases like "host-passthrough" to get the same CPU as the node
+	// and "host-model" to get CPU closest to the node one.
+	// Defaults to host-model.
+	// +optional
+	Model string `json:"model,omitempty"`
+
+	// DedicatedCPUPlacement requests the scheduler to place the VirtualMachineInstance on a node
+	// with enough dedicated pCPUs and pin the vCPUs to it.
+	// +optional
+	DedicatedCPUPlacement bool `json:"dedicatedCPUPlacement,omitempty"`
+
+	// NUMA allows specifying settings for the guest NUMA topology
+	// +optional
+	NUMA *v1.NUMA `json:"numa,omitempty"`
+
+	// IsolateEmulatorThread requests one more dedicated pCPU to be allocated for the VMI to place
+	// the emulator thread on it.
+	// +optional
+	IsolateEmulatorThread bool `json:"isolateEmulatorThread,omitempty"`
```

Comment:
I wonder if it would be safer to use a struct here in order to keep the door open to reflect different emulator thread isolation policies in future.
@vladikr thoughts?

Reply:
OTOH - Maybe we should stick aligned to the VMI API, as we'd need the change there as well.